### ISSUE TITLE
How to integrate with PM2?

### ISSUE THREAD

**Issue by jeghers** [wellyshen/react-cool-starter#98]
Need a good way to run node production mode as a system demon on a Linux server. PM2 looks like a very good way to do this, can you help me figure out and document how to use PM2 in order to make this a standard demon service on a Linux machine?

**Comment by martin2786** [comment]
I may be able to help you with this. Send me an email and we can exchange ideas.

**Comment by wellyshen** [comment]
@martin2786 Thank you for handling this issue 👍

**Comment by jeghers** [comment]
Gmail will not let me reply directly to Martin, so here is my email...
mark.jeghers@gmail.com

**Comment by martin2786** [comment]
Good, I'll get in touch with you shortly.

**Comment by jeghers** [comment]
Cool thanx

**Comment by elrumordelaluz** [comment]
Hi y'all, just curious how you handled the zero-downtime with `pm2`. 
Did you `yarn build` in the server or not ignore the `/build` folder to commit in the repo?
Also how you handle `env` variables to avoid commit in the repo.

In case my email is elrumordelaluz [@] hotmail [.] com

thanks in advance!

**Comment by martin2786** [comment]
Hi @elrumordelaluz to run the starter with pm2 you need to do the following:

- Install pm2 globally: npm i pm2 -g 
- on the starter folder create a pm2 ecosystem (this will contain all the config for pm2) with: pm2 ecosystem
- change the ecosystem file created to contain something like this:
module.exports = {
  apps: [
    // First application
    {
      name: 'cool_starter2',
      script: 'index.js',
      env: {
        NODE_PATH: './src',
        PORT: 3000,
        NODE_ENV: 'production',
      },
      env_production: {
        NODE_ENV: 'production',
      },
    }],
};
- run pm2 with the ecosystem file: pm2 start ecosystem.config.js

This is the basic config for pm2, take a look at pm2 docs to see what else can you add (max memory usage, cluster, etc)

About how to handle env variables, I usually go with this: https://github.com/motdotla/dotenv
but there are several solutions.

Cheers.

**Comment by martin2786** [comment]
Also it's usually a good practice to have a default env file in the repo with the basic config (no keys or any sensitive data). And then keep a local file in each environment that contains all the secret info. Not all env values are sensitive or need to be hidden.

**Comment by jeghers** [comment]
This looks useful, I gotta try this.  I assume "ecosystem" can be named
something else, like "service" or "daemon", yes?

**Comment by martin2786** [comment]
Actually the ecosystem file can not be renamed because otherwise the CLI tool won't read the file as an ecosystem file. Think of the ecosystem as the definition of a daemon or a service, it has all the config for that daemon to start, even the name of the daemon will be the one defined as Name in the ecosystem file.

Each time you run a different ecosystem it will create a new daemon, for example:
- if you run pm2 start ecosystem.config.js and the name is react_starter it will daemonize that node app and start it with the name react_starter. if you run again the same file but you change the name to cool_starter it will create a different daemon with that new name.

![image](https://user-images.githubusercontent.com/2111808/30502635-7bfc305a-9a3d-11e7-9bdd-2e4395abd063.png)

In that image I've used the same ecosystem file but with a different name so it has created 2 different services. It is the same application, running as two different processes on 2 different ports.

So to summarize, remember that anytime you run a new ecosystem.config.js file, it will create a daemon with that configuration.

**Comment by elrumordelaluz** [comment]
hola @martin2786 thanks for the reply!
I am doing similar steps as you describe, below I'll point the specific stuff I am asking:

### env
Regarding the `env` variables I solved whit a local file in production server (same as you proposed), calling it in `post-deploy` form `ecosystem.json`. Asked because is an [open issue](https://github.com/Unitech/pm2-deploy/issues/37) and wanted to know your solution.

### build
> Did you yarn build in the server… 
I asked about the deployment process, more interested in how you handle the `build` process. Since this _boilerplate_ has a `build` script to made the code _production ready_ I am curious if 
- you run this script in your local machine, so commint/push the complete `/build` folder to your repo or
- you run this scrip in a _pre-deploy_ hook, so only have the `/src` files in the repo and build in the server
In one project I tried both and found the last one has issues when the script clean the `build` folder where `pm2` is looknig for. Any of you have the same problem? Are there other options/solutions?

Thanks again,

**Comment by martin2786** [comment]
Hi @elrumordelaluz 
## Env
I think you went with a good solution there. 

## Build
- I think it depends on your deployment process, I've always used a CI  solution (jenkins, teamcity, bamboo) so there was always a step that run the build and generated the needed files before deploying. It takes time to get a smooth workflow because there are usually problems with cleaning folders or files, so don't feel alone on that issue. I found that building a docker image and deploying the image can solve all those problems, but it's a whole new world that needs time to learn and implement.

**Comment by elrumordelaluz** [comment]
yup! @martin2786 gracias

**Comment by martin2786** [comment]
@jeghers is it ok if this issue gets closed? do you need anything else regarding PM2?

**Comment by jeghers** [comment]
Ya I guess.  Still gotta plug the PM2 process into a :Linux daemon, but
there are articles on that...

**Comment by martin2786** [comment]
@jeghers I think I misunderstood your question, but, pm2 is the daemon, what do you mean by plugging it into a linux daemon? if you start pm2 in the linux OS, it will start a daemon so I don't think you really need to plug anything else to it.

**Comment by jeghers** [comment]
I mean integrating into systemd so it starts at bootup and can be
controlled (start/stop/status) like the other services.

**Comment by martin2786** [comment]
That's different, then I think you might need to look at this http://pm2.keymetrics.io/docs/usage/startup/

**Comment by elrumordelaluz** [comment]
probably `pm2 startup systemd`

**Comment by wellyshen** [comment]
@jeghers Can you close this issue if you got the help?

**Comment by jeghers** [comment]
Still pending, I have to look into it more this weekend, stand by please...

**Comment by wellyshen** [comment]
@jeghers Can it be closed?

**Comment by jeghers** [comment]
Yes

**Comment by redlinecodes** [comment]
I don't mean to reopen the issue, but is all this still relevant?  I literally just installed pm2 globally on my local machine and then went into my local folder for this repo and ran the command pm2 start index.js and it just started right up... ??

**Comment by wellyshen** [comment]
@redlinecodes Don't forget the necessary process.env, for example:

```
module.exports = {
  apps: [
    {
      // other settings...
      script: 'index.js',
      env: {
        NODE_ENV: 'production'
      }
    }
  ]
};

```
